src/main.cpp

```diff
@@ -182,30 +186,38 @@ void main_preinit() {
     //__HAL_RCC_GET_FLAG(RCC_FLAG_BORRST);
     __HAL_RCC_CLEAR_RESET_FLAGS();
 
-    /* USER CODE END 1 */
-
-    /* MCU Configuration--------------------------------------------------------*/
-
     /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
-    HAL_Init();
-
-    /* USER CODE BEGIN Init */
-
-    /* USER CODE END Init */
-
+    HAL_Init(); //it is low level enought to be run in startup script
     /* Configure the system clock */
     SystemClock_Config();
 
-    /* USER CODE BEGIN SysInit */
-
     diag_check_fastboot();
 
-    /* USER CODE END SysInit */
+    MX_I2C1_Init();
+    crc32_init();
+
+    int irq = __get_PRIMASK() & 1;
+    __enable_irq();
 
+    eeprom_init_status_t status = eeprom_init();
+    if (status == EEPROM_INIT_Defaults || status == EEPROM_INIT_Upgraded) {
+        // this means we are either starting from defaults or after a FW upgrade -> invalidate the XFLASH dump, since it is not relevant anymore
+        dump_in_xflash_reset();
```

Comment:
have you checked that `dump_in_xflash_reset` is low-level enough to be left here? I think the SPI XFLASH is being inited in main(), which is AFTER `EepromSystemInit`.

Reply:
it was initialized internally in this function too, but you are right it is better to do it in main